### PR TITLE
SPV_QCOM_image_processing support

### PR DESCRIPTION
**Pull request by wooyoungqcom**
Adding support for SPV_QCOM_image_processing

### CODE REVIEW

**Review by alan-baker — CHANGES_REQUESTED** [review]
Please add tests to cover this new functionality.

**Comment by alan-baker** [comment]
Take a look at `test/val/val_image_test.cpp`. That contains the other image validation tests.

File: source/val/validate_image.cpp [hunk 1]
```diff
@@ -2249,12 +2322,106 @@ spv_result_t ImagePass(ValidationState_t& _, const Instruction* inst) {
...
+      return ValidateImageProcessingQCOM(_, inst);
+
     default:
       break;
   }
 
   return SPV_SUCCESS;
 }
 
+bool IsImageInstruction(const spv::Op opcode) {
```
Comment (spencer-lunarg) [inline source/val/validate_image.cpp:2338]: Is this not something we have anyway of generating currently? Do you really care for `OpImageSparseTexelsResident` which doesn't have a `OpImage` ID to it?

Comment (wooyoungqcom) [reply source/val/validate_image.cpp:2338]: Thanks for the comment. I was not very sure about OpImageSparseTexelsResident, so I just added it. Let me remove it.

File: source/val/validation_state.h [hunk 2]
```diff
@@ -820,6 +833,10 @@ class ValidationState_t {
   std::unordered_map<uint32_t, std::vector<Instruction*>>
       sampled_image_consumers_;
 
+  /// Stores load instructions that load texturesed used
```
Comment (alan-baker) [inline source/val/validation_state.h:836]:
```suggestion
  /// Stores load instructions that load texturese used
```

File: source/val/validation_state.h [hunk 3]
```diff
@@ -820,6 +833,10 @@ class ValidationState_t {
...
+  //  in QCOM image processing functions
+  std::unordered_set<uint32_t> tpip_texture_consumers_;
```
Comment (alan-baker) [inline source/val/validation_state.h:838]: I'm not sure this is a good variable name, what is `tpip`? Maybe `qcom_image_processing_consumers` would be better.

File: source/val/validate_image.cpp [hunk 4]
```diff
@@ -1087,6 +1091,23 @@ spv_result_t ValidateSampledImage(ValidationState_t& _,
...
+    const spv_parsed_operand_t& topnd = inst->operand(2);
+    int t_idx = (int)inst->word(topnd.offset);
+    ld_inst = _.FindDef(t_idx);
+  }
+
+  if (ld_inst->opcode() == spv::Op::OpLoad) {
+    int texture_id;
+    const spv_parsed_operand_t& popnd =
+        ld_inst->operand(2);  // variable to load
+    texture_id = (int)ld_inst->word(popnd.offset);
```
Comment (alan-baker) [inline source/val/validate_image.cpp:1106]: Prefer not mixing operand and word calls.

File: source/val/validate_image.cpp [hunk 5]
```diff
@@ -2130,6 +2151,58 @@ spv_result_t ValidateImageSparseTexelsResident(ValidationState_t& _,
...
+    case spv::Op::OpImageBlockMatchSADQCOM: {
+      const spv_parsed_operand_t& topnd = inst->operand(2);  // target
+      int tgt_idx = (int)inst->word(topnd.offset);
+      res = ValidateImageProcessingQCOMDecoration(
+          _, tgt_idx, spv::Decoration::BlockMatchTextureQCOM);
+      const spv_parsed_operand_t& ropnd = inst->operand(4);  // reference
+      int ref_idx = (int)inst->word(ropnd.offset);
+      spv_result_t res1 = ValidateImageProcessingQCOMDecoration(
+          _, ref_idx, spv::Decoration::BlockMatchTextureQCOM);
+      res = (res != SPV_SUCCESS ? res : res1);
```
Comment (alan-baker) [inline source/val/validate_image.cpp:2196]: Why not just return earlier if `res` is not success?

**Review by alan-baker — CHANGES_REQUESTED** [review]

File: source/val/validation_state.h (hunk 2 above)
Comment (wooyoungqcom) [reply source/val/validation_state.h:836]: Fixed.

File: source/val/validate_image.cpp [hunk 6]
```diff
@@ -2130,6 +2151,58 @@ spv_result_t ValidateImageSparseTexelsResident(ValidationState_t& _,
...
+    si_inst = ld_inst;
+    const spv_parsed_operand_t& topnd = si_inst->operand(2);  // texture
+    int t_idx = (int)si_inst->word(topnd.offset);
+    ld_inst = _.FindDef(t_idx);
+  }
+  assert(ld_inst->opcode() == spv::Op::OpLoad);
+  const spv_parsed_operand_t& popnd = ld_inst->operand(2);  // variable to load
+  int texture_id = (int)ld_inst->word(popnd.offset);
+  if (!_.HasDecoration(texture_id, decor)) {
+    return _.diag(SPV_ERROR_INVALID_DATA, ld_inst) << "Missing decoration";
```
[the remark this answers is not in the document]
Comment (wooyoungqcom) [reply source/val/validate_image.cpp:2167]: Fixed.

File: source/val/validation_state.h (hunk 3 above)
Comment (wooyoungqcom) [reply source/val/validation_state.h:838]: Done

File: source/val/validate_image.cpp (hunk 5 above)
Comment (wooyoungqcom) [reply source/val/validate_image.cpp:2195]: Done.

File: source/val/validate_image.cpp (hunk 4 above)
Comment (wooyoungqcom) [reply source/val/validate_image.cpp:1105]: Could you suggest the SPIRV-Tool's preferred way of accessing an operand's ID ?

Comment (wooyoungqcom) [reply source/val/validate_image.cpp]: Thanks much, Alan. Fixed.

File: source/val/validate_image.cpp [hunk 7]
```diff
@@ -2130,6 +2151,58 @@ spv_result_t ValidateImageSparseTexelsResident(ValidationState_t& _,
...
+                                                   spv::Decoration decor) {
+  const Instruction* si_inst = nullptr;
+  const Instruction* ld_inst = _.FindDef(id);
+  if (ld_inst->opcode() == spv::Op::OpSampledImage) {
+    si_inst = ld_inst;
+    const spv_parsed_operand_t& topnd = si_inst->operand(2);  // texture
+    int t_idx = (int)si_inst->word(topnd.offset);
+    ld_inst = _.FindDef(t_idx);
+  }
+  assert(ld_inst->opcode() == spv::Op::OpLoad);
```
[the remark this answers is not in the document]
Comment (wooyoungqcom) [reply source/val/validate_image.cpp:2157]: Good point. Thanks for catching it. We will make changes to allow for an array of textures.

[the remark this answers is not in the document]
Comment (wooyoungqcom) [reply source/val/validate_image.cpp:2157]: Just realized that even if an array is used, the element texture needs to be loaded before passed to the intrinsic instruction. Do you agree ? If you do, turning the assertion to an error would be good enough ? I will try to update the test case.

[the remark this answers is not in the document]
Comment (alan-baker) [reply source/val/validate_image.cpp:2157]: With an arrayed texture you'd have a setup like the following:
```
%weight_texture = OpTypeImage ...
%sampler = OpTypeSampler
%sampled_image = OpTypeSampledImage %weight_texture
%array = OpTypeArray %weight_texture ...size...
%ptr_array = OpTypePointer UniformConstant %array
%ptr_texture = OpTypePointer UniformConstant %weight_texture
%var = OpVariable %ptr_array UniformConstant
...
%access_chain = OpAccessChain %ptr_texture %var %index
%load_tex = OpLoad %weight_texture %access_chain
%load_s = OpLoad %sampler ...
%sampled = OpSampledImage %sampled_image %load_tex %load_s
```

So the concern is about tracing back to the OpVariable more than the OpLoad. I think the OpLoad should be an error if there are no other possibilities. For the subsequent check of the variable decoration would go awry in the above case.

[the remark this answers is not in the document]
Comment (wooyoungqcom) [reply source/val/validate_image.cpp:2157]: We apply decoration BlockMatchTextureQCOM/WeightTextureQCOM to "textures used as Target Sampled Image and Reference Sampled Image in OpImageBlockMatchSSDQCOM/OpImageBlockMatchSADQCOM / OpImageSampleWeightedQCOM". )

So, I think we don't need to track back to OpVariable which is a pointer variable if an array is used.

```
         OpDecorate %53 BlockMatchTextureQCOM
         ...
         %42 = OpTypeImage %float 2D 0 0 0 1 Unknown
         %43 = OpTypeSampledImage %42
         ...
%_arr_43_uint_3 = OpTypeArray %43 %uint_3
%_ptr_UniformConstant__arr_43_uint_3 = OpTypePointer UniformConstant %_arr_43_uint_3
   %samplers = OpVariable %_ptr_UniformConstant__arr_43_uint_3 UniformConstant
         ...
         %53 = OpAccessChain %_ptr_UniformConstant_43 %samplers %int_1
         %54 = OpLoad %43 %53
         %55 = OpLoad %v2uint %ref_coords
         %56 = OpLoad %v2uint %blockSize
         %57 = OpImageBlockMatchSADQCOM %v4float %46 %47 %54 %55 %56
```

**Review by alan-baker — APPROVED** [review]

**Comment by alan-baker** [comment]
Looks like you just need to fix the formatting.

**Comment by wooyoungqcom** [comment]
Thanks much, Alan. I just uploaded a new change set with the format fixes.

**Comment by alan-baker** [comment]
There are still some formatting errors in val_image_test.cpp. If in doubt, you can copy the diff into a file and apply it as a patch to make the format bot happy.

**Comment by wooyoungqcom** [comment]
done. (sorry I missed val_image_test.cpp)

**Pull request merged**